fairscale/nn/data_parallel/sharded_ddp.py

```diff
@@ -4,234 +4,266 @@
+
+    Keyword Args:
+        process_group (torch.nn.Optimizer):
+            optimizer to shard (default: SGD)
+        process_group (group):
+            torch.distributed group (default: group.WORLD)
+        broadcast_buffers (bool):
+            whether to broadcast model buffers in between ranks at the beginning of each forward pass
+        reduce_buffer_size (int):
+            the size of the per-device-per-rank buffers used for the reduce operation
+
     """
 
     def __init__(
         self,
-        module: nn.Module,
-        optimizer: Type[torch.optim.Optimizer],
-        optimizer_params: Dict[str, Any],
-        world_size: int,
-        broadcast_buffers: bool,
+        base_model: nn.Module,
+        sharded_optimizer: Union[OSS, List[OSS]],
         process_group: Any = None,
-        buffer_size: int = 2 ** 19,
+        broadcast_buffers: bool = True,
     ):
         super().__init__()
 
-        self.module = module
-        self.world_size = world_size
+        self.base_model = base_model
+        self.sharded_optimizers = [sharded_optimizer] if isinstance(sharded_optimizer, OSS) else sharded_optimizer
+        self.enable_broadcast_buffers = broadcast_buffers
+        self.accumulate_grads = False
+
+        # Communication related attributes
         self.process_group = process_group if process_group is not None else dist.group.WORLD
+        self.world_size = dist.get_world_size(self.process_group)
+        self.reference_global_rank = OSS.get_global_rank(self.process_group, 0)  # picking rank 0 as the reference
         self.rank = dist.get_rank(self.process_group)
-        self.broadcast_buffers = broadcast_buffers
-        self.authoritative_rank = 0
+        self.global_rank = OSS.get_global_rank(self.process_group, self.rank)
+
+        # Expose the same attributes as PytorchDDP, some frameworks rely on them.
+        # See https://pytorch.org/docs/stable/_modules/torch/nn/parallel/distributed.html#DistributedDataParallel
+        self.devices = {p.device for p in self.base_model.parameters()}
+        self.is_multi_device_module = len(self.devices) > 1
+        distinct_device_types = {p.device.type for p in self.base_model.parameters()}
+        assert len(distinct_device_types) == 1, (
+            "ShardedDataParallel's input module must be on "
+            "the same type of devices, but input module parameters are located on {} different device types."
+        ).format(distinct_device_types)
+        self.device_type = list(distinct_device_types)[0]
+
+        # Scafolding to be able to reduce the grads during the BW pass
+        self._param_iterator = chain(*[optim.param_bucket_strategy.keys() for optim in self.sharded_optimizers])
+        self._grad_to_be_reduced = [True for _ in self._param_iterator]
+        self._grad_accs: List[Callable] = []
+        self._setup_backward_hooks()
+
+        # Make sure that all ranks start with the same model
```

Comment:
Why wouldn't they?

Reply:
Discussed over chat - this is just in beginning of training in case there are different seeds.
Small optimization could be to make this optional. For example, if we're loading from a saved checkpoint, we can assume that the parameters are already synced. For large models, this can save a minute of time and make startup faster.

Reply:
done, now optional